### ISSUE TITLE
CloudTable reference in Azure Table Storage is a singleton but it is not threadsafe

### ISSUE THREAD

**Issue by rmja** [dotnet/orleans#5003]
The `AzureTableDataManager` used by `AzureTableGrainStorage` which is registered as a singleton has a single instance of `CloudTable`, but this type [is not thread safe](https://stackoverflow.com/questions/13246720/is-azure-cloudtable-thread-safe). This results in exceptions thrown by the AzureStorage under load from `AzureTableDataManager.ReadSingleTableEntryAsync`:

```
Microsoft.WindowsAzure.Storage.StorageException: An item with the same key has already been added. Key: timeout ---> System.ArgumentException: An item with the same key has already been added. Key: timeout
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at Microsoft.WindowsAzure.Storage.Shared.Protocol.HttpRequestMessageFactory.CreateRequestMessage(HttpMethod method, Uri uri, Nullable`1 timeout, UriQueryBuilder builder, HttpContent content, OperationContext operationContext, ICanonicalizer canonicalizer, StorageCredentials credentials)
   at Microsoft.WindowsAzure.Storage.Table.Protocol.TableOperationHttpRequestMessageFactory.BuildRequestCore(Uri uri, UriQueryBuilder builder, HttpMethod method, Nullable`1 timeout, HttpContent content, OperationContext ctx, ICanonicalizer canonicalizer, StorageCredentials credentials)
   at Microsoft.WindowsAzure.Storage.Table.Protocol.TableOperationHttpRequestMessageFactory.BuildRequestForTableQuery(Uri uri, UriQueryBuilder builder, Nullable`1 timeout, HttpContent content, OperationContext ctx, TablePayloadFormat payloadFormat, ICanonicalizer canonicalizer, StorageCredentials credentials)
   at Microsoft.WindowsAzure.Storage.Table.TableQuery`1.<>c__DisplayClass24_0`2.<QueryImpl>b__0(RESTCommand`1 cmd, Uri uri, UriQueryBuilder queryBuilder, HttpContent cnt, Nullable`1 serverTimeout, OperationContext ctx)
   at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.ExecuteAsyncInternal[T](RESTCommand`1 cmd, IRetryPolicy policy, OperationContext operationContext, CancellationToken token)
   --- End of inner exception stack trace ---
   at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.ExecuteAsyncInternal[T](RESTCommand`1 cmd, IRetryPolicy policy, OperationContext operationContext, CancellationToken token)
   at Microsoft.WindowsAzure.Storage.Table.TableQuery`1.<>c__DisplayClass20_0.<<ExecuteQuerySegmentedAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Orleans.Persistence.AzureStorage.AzureTableDataManager`1.ReadSingleTableEntryAsync(String partitionKey, String rowKey)
   at SampleStore.Silo.CustomAzureTableGrainStorage.GrainStateTableDataManager.Read(String partitionKey, String rowKey) in C:\Source\Utiliread\src\SampleStore.Silo\CustomAzureTableGrainStorage.cs:line 408
   at SampleStore.Silo.CustomAzureTableGrainStorage.ReadStateAsync(String grainType, GrainReference grainReference, IGrainState grainState) in C:\Source\Utiliread\src\SampleStore.Silo\CustomAzureTableGrainStorage.cs:line 74
   at Orleans.Core.StateStorageBridge`1.ReadStateAsync()

```

The `Table` property on `AzureTableDataManager` is unfortunately public, so I don't see an easy fix (that I can do). If it was private to the manager, one could simply have a `ConcurrentBag<CloudTable>` where one can borrow initialized clients from.

**Comment by rmja** [comment]
It turns out that I get this error when I am activating too many cold grains from table storage in a `Task.WhenAll()`. The `ExecuteQuerySegmentedAsync` call then throws an exception with status code 306 which is uncaught.

**Comment by rmja** [comment]
I think this is very related to what I am seeing: https://github.com/Azure/azure-storage-net/issues/620

**Comment by vmendi** [comment]
We are also seeing this exception, but not even under a lot of load.

**Comment by ReubenBond** [comment]
@rmja sounds correct. I saw that one of our internal users also hit the same issue. I took a stab at implementing an easy fix, but I'm I don't know how to build/test that library.

https://github.com/Azure/azure-storage-net/compare/master...ReubenBond:fix-620?expand=1

**Comment by rmja** [comment]
@ReubenBond I sent a PR https://github.com/Azure/azure-storage-net/pull/774. It seems to fix the issue.

**Comment by ReubenBond** [comment]
Thanks, @rmja - your fix looks good, much better than mine 😜

**Comment by rmja** [comment]
@ReubenBond It seems as if the Table storage API is not supported by the Azure Storage API anymore: https://github.com/Azure/azure-storage-net#features This seems akward! - but how do we proceed from here?

**Comment by rmja** [comment]
I created this issue, lets see what they have to say: https://github.com/Azure/azure-storage-net/issues/779

**Comment by ReubenBond** [comment]
Interesting turn of events 😅 - thanks for following up with them. I suppose we need to wait for this new .NET Standard API to become available, as [mentioned in the linked issue](https://github.com/Azure/azure-storage-net/issues/779#issuecomment-426722826).

**Comment by ilyalukyanov** [comment]
@rmja @ReubenBond I have a naive idea. Would something like this do the trick? https://github.com/dotnet/orleans/compare/master...ilyalukyanov:fix-5003?expand=1

**Comment by rmja** [comment]
The thing is that I am under the impression that the root of the exception in the description is a problem in the Azure Storage SDK, a problem for which [I created a patch which probably never will be merged...](https://github.com/Azure/azure-storage-net/pull/774#issuecomment-426652496). I am currently running with a custom build of the SDK with the patch applied, and I havn't seen any issues with it.

@ilyalukyanov I have tried a similar approach as you suggest, but it it didn't resolve the issue for me.

I think we are left in a vacuum and will have to wait until the new CosmosDB SDK with Table support is released before we can proceed with this.
